kuebiko/pipeline/sic/flow.py

```diff
@@ -0,0 +1,146 @@
+        ## the overhead of persisting artifacts.
+        self.url = URL
+
+        ## Naming an artifact with a leading underscore makes the data artifact
+        ## available to future steps as usual but hides it from a user inspecting
+        ## the results of the run.
+        ## This is useful to avoid users loading the raw data rather than the processed
+        ## data, but it makes it harder to debug/monitor artifacts so it's a case
+        ## of best judgement.
+        self._raw_data = excel_to_df(get(self.url))
+
+        self.next(self.transform)
+
+    ## What operations should belong to the same step and when does it make
+    ## sense to split a large step into multiple separate steps?<br>
+    ## Artifacts are persisted when a step finishes executing. After artifacts
+    ## have been persisted successfully, they become available for inspection,
+    ## e.g. `metaflow.Run("Sic2007Structure").start.task.data.url`.
+    ## Also you will be available to
+    ## [resume](https://docs.metaflow.org/metaflow/
+    ## debugging#how-to-use-the-resume-command)
+    ## execution at an arbitrary step.
+    ## Hence keeping steps reasonably small in terms of execution time,
+    ## means you won't lose much work if failures happen; however persisting
+    ## artifacts (particularly using the S3 datastore) and launching tasks
+    ## (particularly using AWS batch) has some overhead so if your steps are
+    ## too small, the overhead starts dominating the total execution time.
+    ## **On balance it is good to have steps that logically map to the domain/problem
+    ## so that it is easy to get a high-level understanding of the flow.** If
+    ## you notice a large amount of overhead or your flow is hard to debug then
+    ## you can split or merge existing steps accordingly.
+    @step
+    def transform(self):
+        """Transform data.
+
+        Make implicit entries explicit at row-level (fill), normalise SIC
+        codes, and add extra codes specific to Companies House.
+        """
+        from utils import companies_house_extras, fill, normalise_codes
+
+        ## One of the big downsides of passing around dataframes is that it's
+        ## easy to lose track of the shape of the data as it passes through
+        ## various transformation functions, especially if you didn't even
+        ## write the code in the first place!<br>
+        ## **Give functions informative names and docstrings and obey the
+        ## Single Responsibility Principle** to really help the understandability
+        ## of code.<br>
+        ## What about these functions, are there names and docstrings
+        ## informative enough? What would you change?<br>
+        ## In a later episode we will see how unit tests can also increase the
+        ## readability and transparency of these transformation functions.
+        self._data = (
+            self._raw_data.pipe(fill)
```

Comment:
I guess `fill` could have a more informative name? ie, fill what (presumably nulls) and with what?
Relatedly, I'm not sure what "Make implicit entries explicit at row-level (fill)" means in the docstring.